### PR TITLE
[tool] Add option for Android compile SDK version to update-dependencies command

### PR DESCRIPTION
**Pull request by camsim99**
Adds option to `update-dependencies` command to update the compile SDK version of plugins or their example apps.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: script/tool/lib/src/update_dependency_command.dart [hunk 1]
```diff
@@ -233,59 +250,110 @@ ${response.httpResponse.body}
...
+        : PackageResult.skip('No example apps run on Android.');
+  }
+
+  Future<PackageResult> _runForCompileSdkVersion(
+      RepositoryPackage package) async {
+    if (!package.platformDirectory(FlutterPlatform.android).existsSync()) {
+      return PackageResult.skip(
+          'Package ${package.displayName} does not run on Android.');
+    } else if (package.getEnclosingPackage() != null) {
+      // We skip examples for this command.
```
Comment (camsim99) [inline script/tool/lib/src/update_dependency_command.dart:334]: Would appreciate if someone can sanity-check me on whether or not this is a valid way to ensure we don't run on examples running on Android.

Comment (stuartmorgan) [reply script/tool/lib/src/update_dependency_command.dart:334]: There are edge cases where it probably isn't exactly what you want; https://github.com/flutter/packages/tree/main/packages/pigeon/platform_tests/test_plugin comes to mind (this is a plugin which in inside a package, but which we probably want automagic updates to work on).

If you want it to specifically apply to examples, you could make a `RepositoryPackage` `isExample` method that calls `getEnclosingPackage`, then calls `getExamples` on that, and then checks if any of them is the package you started with.

File: script/tool/lib/src/common/repository_package.dart [hunk 2]
```diff
@@ -142,6 +142,23 @@ class RepositoryPackage {
...
+      // An example package is enclosed in another package.
+      return false;
+    }
+    for (final RepositoryPackage example in enclosingPackage.getExamples()) {
+      if (example.path == path) {
+        // This package is an example of its enclosing package.
+        return true;
+      }
+    }
+    return false;
```
Comment (stuartmorgan) [inline script/tool/lib/src/common/repository_package.dart:159]: Optional: Everything from 153 to here could collapse to:
```dart
// Check whether this is one of the enclosing package's examples.
return enclosingPackage.getExamples().any((RepositoryPackage p) => p.path == path);
```

File: script/tool/lib/src/update_dependency_command.dart [hunk 3]
```diff
@@ -42,10 +42,17 @@ class UpdateDependencyCommand extends PackageLoopingCommand {
...
+          _AndroidDepdencyType.compileSdkForExamples,
         ],
         allowedHelp: <String, String>{
-          'gradle': 'Updates Gradle version used in plugin example apps.',
+          _AndroidDepdencyType.gradle:
+              'Updates Gradle version used in plugin example apps.',
+          _AndroidDepdencyType.compileSdk:
+              'Updates compileSdk version used to compile plugins.',
+          _AndroidDepdencyType.compileSdkForExamples:
+              'Updates compileSdk version used to compile plugin examples',
```
Comment (stuartmorgan) [inline script/tool/lib/src/update_dependency_command.dart:55]: Nit: This is missing the `.` that the others have.

File: script/tool/lib/src/update_dependency_command.dart [hunk 4]
```diff
@@ -233,59 +250,110 @@ ${response.httpResponse.body}
...
-        updateRanForExamples = true;
-        Directory gradleWrapperPropertiesDirectory =
-            example.platformDirectory(FlutterPlatform.android);
-        if (gradleWrapperPropertiesDirectory
+    return PackageResult.fail(<String>[
+      'Target Android dependency $_androidDependency is unrecognized.'
+    ]);
+  }
+
+  Future<PackageResult> _runForAndroidDependencyOnExample(
```
Comment (stuartmorgan) [inline script/tool/lib/src/update_dependency_command.dart:266]: Nit: `...Examples`

File: script/tool/lib/src/update_dependency_command.dart [hunk 5]
```diff
@@ -233,59 +250,110 @@ ${response.httpResponse.body}
...
+      if (!example.platformDirectory(FlutterPlatform.android).existsSync()) {
+        continue;
+      }
+
+      updateRanForExamples = true;
+      Directory androidDirectory =
+          example.platformDirectory(FlutterPlatform.android);
+      File? fileToUpdate;
+      RegExp? dependencyVersionPattern;
+      String? newDependencyVersionEntry;
```
Comment (stuartmorgan) [inline script/tool/lib/src/update_dependency_command.dart:280]: These should be `final` and non-nullable, and the `if`/`else if` chain should have an `else` that throws. Then you won't need to force-unwrap these later, and the compiler will enforce that they are set in every branch.

**Review by stuartmorgan — APPROVED** [review]
LGTM with minor changes.

File: script/tool/lib/src/common/repository_package.dart (hunk 2 above)
Comment (camsim99) [reply script/tool/lib/src/common/repository_package.dart:159]: Oooo this is neat!

File: script/tool/lib/src/update_dependency_command.dart (hunk 5 above)
Comment (camsim99) [reply script/tool/lib/src/update_dependency_command.dart:280]: This is actually super good to know, thank you!!

**Pull request merged**